### ISSUE TITLE
LCD initialization/usage have high failure rate compare to 0.91 version

### ISSUE THREAD

**Issue by luc-github** [luc-github/Repetier-Firmware-0.92#10]
At initialization the AiO boards have very high failure rate and get white screen or 2 lines visibles
When printing screen become white randomly and need reboot on 1.0 old generation.

**Comment by luc-github** [comment]
Interresting, actually you reduced several timing when we increase them - your experienced improvement by decreasing timing ?
I see also you send clear screen command before initialization is finished - any reason for this ?

I will try your sequence

**Comment by garry572** [comment]
yes according to the data sheets I have, most command complete in a max of 120 usec apart from the clear screen which is 4.0 msec The second clear screen was to see if the screen would recover with being initialized twice I have restarted my 1.0 quite a few time with a single failure, even on the first clear screen. I think part of the issue is that during initialization, the IO line goes from non programmed Io to a digital output, I think sometime the screen sees the Enable line toggled from high to low and take it as a beginning of a command

**Comment by luc-github** [comment]
I have applied your suggestion timing/sequence https://github.com/luc-github/Repetier-Firmware-0.92/commit/d653f5c80764a463a5b487ec7ecce7fb27e2fd7f

**Comment by garry572** [comment]
hi luc hope it helps the weather here has taken a turn for the worse, next week will not be under 30C 
Ive had not LCD initialization failure since I made  the change and since it only show up at low temps I will have to wait a while

**Comment by luc-github** [comment]
I just got white screen just after applied your modifications and another modification present on 0.91 for the write function, so I removed the 0.91 write modification and was fine.
I am not sure the other modification was the root cause of the problem, I did not double check by adding modification back to confirm, I may do later - anyway few change step by step is better 
thanks again for your help and suggestions

Now need to wait someone test on 1.0A to see if there is some improvement for them

**Comment by garry572** [comment]
that's interesting, i will try the write modification on my 1.0 to see it make any changes good or bad

**Comment by luc-github** [comment]
ok - I am doing some printing test now - need few hours before I can reflash printer

**Comment by garry572** [comment]
I'm not sure if im looking in the right place for the 0.91 changes if I'm looking correctly the changes effect the method of setting delays

**Comment by luc-github** [comment]
yes was a delay line 482
DELAY1MICROSECOND;

}
void initializeLCD()

I changed to 

#if CPU_ARCH == ARCH_AVR
    __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");
#else
    HAL::delayMicroseconds(100);
#endif
}
void initializeLCD()

then put DELAY1MICROSECOND; back

**Comment by garry572** [comment]
interesting I will look at it tomorrow, I have a arduino due i will do some testing/verification to see if the delays are generated as expected

**Comment by luc-github** [comment]
Ok thanks a lot for your help ^_^

**Comment by luc-github** [comment]
Interresting, Repetier did some change for " faster lcd timing" by ... increasing some timing actually  (DELAY2MICROSECOND;  and HAL::delayMicroseconds(40);) and reducing the UI_DELAYPERCHAR from 320 to 40 in lcdWriteNibble and lcdWriteByte

https://github.com/repetier/Repetier-Firmware/blob/bd45683f32aa4ba442d41f922f5da16a5e608a51/src/ArduinoDUE/Repetier/ui.cpp

What do you think ?

**Comment by garry572** [comment]
Thats kind of odd, I have a few data sheets comparing supposed compatible  HD44870 and they are all different. I have on showing command delays as short as 32uSec and other as long as 150uSec. Ive also now found one from Hitachi that's a later version of the I have, and the initialization sequence for 4 bit display interfaces is long with 2 additional commands

**Comment by luc-github** [comment]
Yes that what I found also when working on 0.91 improvement - oscahie and I did several test increasing or adding timing, changing sequence, even using other library
So I have requested some help to have more tester/feedback on change we could do - http://voltivo.com/forum/davinci-firmware/540-lcd-failure-on-1-0-1-0a-tester-feedback-needed
Hope it could help to triger the solution - as unfortunatly I think it will be several try and error loops before we get accurate solution and may be solution will be different according the boards

**Comment by garry572** [comment]
That's so true so many possibilities of displays. Mine hasn't faulted since I did the mod so I cant verify more than saying is ok by me.
I will add the latest info to test on my 1.0 but I think other might need to try it out also

**Comment by luc-github** [comment]
Yes you are in same situation as me now ^_^

**Comment by garry572** [comment]
I think it ok to change the UI_DELAYPERCHAR from 320 to 150 as this covers most of the controller variants I have found, anything lower ie 40 might be ok for a fast controller, but theses ones seem slow so it better to be safe

**Comment by superdavex** [comment]
Just saw your request for 1.0a users on voltivo.  I am running .92 compiled from the latest master on a 1.0a.   The two issues I have are the LCD initialization failure and the extruder temperature error.  It usually takes me 2 to 3 times to get the LCD to initialize properly when powering on.  Usually if I turn it on, wait a few seconds, and then turn it off/ on it works.

**Comment by luc-github** [comment]
Hi superdavex, welcome, thanks for your feedback, appreciate
Ok so I will create a branch for testing some change for the LCD, so will be easy to switch to original if it is not working 
for the temperature we will adjust table to avoid the problem - what temperature the bed display and extruder display on repetier when you connect repetier host ?
Also homing Z is working on your 1.0a ? ( I just saw someone raise issue on forum)
What it your time zone ? (this is to know when I can do change to deliver test version to you)

**Comment by superdavex** [comment]
Np, thanks for all the work you have done.  This firmware prints much better than stock.  The bed is -12 to -16 and the extruder is 19.5 to 19.7.  Homing Z works fine for me, I have done many prints with it and never seen an issue.  I am in the US Eastern Standard (UTC−05:00).  I have noticed the LCD issue only occurs when powering off/on from the switch.  If i reboot using left arrow, down arrow, and ok it fixes it every time on the first try.

**Comment by luc-github** [comment]
Sorry can you confirm the version of the FW ? it is in settings - version (1.1-201.....
because you said "extruder temperature error" and now seems the bed is a problem 

that is very interresting that the emergency stop is better than the switch because looks like the timing
````
    // SEE PAGE 45/46 FOR INITIALIZATION SPECIFICATION!
    // according to datasheet, we need at least 40ms after power rises above 2.7V
    // before sending commands. Arduino can turn on way before 4.5V.
    // is this delay long enough for all cases??
    HAL::delayMilliseconds(400);
````

is not enough

I will prepare a branch for testing after my breakfast ^_^

**Comment by superdavex** [comment]
I had those values backwards at first, not sure if you saw the edit.  It is the extruder that is in negative values.  Currently I have the MIN_DEFECT_TEMPERATURE set to -20 as a workaround.

The version is 
#define VERSION_MAJOR " 1"
#define VERSION_MINOR_YEAR "15"
#define VERSION_MINOR_MONTH "01"
#define VERSION_MINOR_DAY "02"
#define VERSION_BUILD "1"

The delay in the master branch was at 500
HAL::delayMilliseconds(500);

I will trying bumping that up.

**Comment by garry572** [comment]
I have had no issues with my LCD since the mods, it likely some boards power up differently. Perhaps someone can confirm what the display is ie what manufacturer and pert number so we can narrow down the roor cause of the failures

**Comment by luc-github** [comment]
I just created a testing branch and changed the value for rise power from 500 to 2000

Ho great - as bed should be OK, table is tweaked already, so I was worry ^_^
Extruder issue is consistant with the current table and my observation of the table data (https://github.com/luc-github/Repetier-Firmware/issues/83) -  I will need ADC value - I will update the testing code to display it on screen so you can give it to me and the temperature displaying for the bed to match them

Garry - I already requested information to the guy who removed his screen when he modified his printer - he said he will check but never got any feedback - removing the screen is not easy on Davinci

**Comment by superdavex** [comment]
Still having the display issues with the 2000ms pause.  I still corrects it every time when using emergency stop.

**Comment by luc-github** [comment]
we have 2 lead : 
Garry suggested to duplicate initialization 
Another lead is repetier did several update recently for LCD 

I will apply the duplicate of initialization on testing branch  first

for extruder temperature I created a  new issue to avoid to mix the threads and already modifed the table in testing branch so you can test and feedback https://github.com/luc-github/Repetier-Firmware-0.92/issues/18

**Comment by luc-github** [comment]
branch updated
https://github.com/luc-github/Repetier-Firmware-0.92/commit/c92b35590542625462a0fadc1f7ef9fb207b9d8f

**Comment by superdavex** [comment]
No change with second initialization.

**Comment by luc-github** [comment]
Ok - I will revert and prepare the second lead : implement all latest repetier update (https://github.com/repetier/Repetier-Firmware/blob/bd45683f32aa4ba442d41f922f5da16a5e608a51/src/ArduinoDUE/Repetier/ui.cpp and https://github.com/repetier/Repetier-Firmware/commit/08d032adca47f0bde63f5bdc13d69eb78e64639d )
I need 1h or more as I need to do something else first - will post when modification is ready

Thanks a lot for your help

**Comment by superdavex** [comment]
Ok,no problem. I'll be around

**Comment by luc-github** [comment]
https://github.com/luc-github/Repetier-Firmware-0.92/commit/e6784ade5a12397d65aeed007b03357da519aa1b ready for test - repetier timing but keep Garry sequence modification

**Comment by superdavex** [comment]
Still about the same frequency of initialization failures.  The LCD has garbled characters on it now instead of black bars though.  Emergency stop still fixes it every time.

**Comment by luc-github** [comment]
Ho before is not white screen but black bars ?

**Comment by superdavex** [comment]
A mix of that and missing text and random characters.  Now its random blank screen and garbled text.

**Comment by luc-github** [comment]
so little bit better as black screen means no init at all
I will put original sequence to see if there is a difference

**Comment by luc-github** [comment]
updated https://github.com/luc-github/Repetier-Firmware-0.92/commit/0a32ed68bd0def0853b08e94317db7185c98e944

**Comment by superdavex** [comment]
Still random garbled text and blank screens.  But now it also has a normal text screen but constantly shifting the location of the text. The failures seem to be more often

**Comment by superdavex** [comment]
Just tried flashing master, definitely more failures compared to master.   With master most failures are blank screens.

**Comment by luc-github** [comment]
ok so back to master = Garry sequence and lets play with timing as we did in 0.91
this screen is terrible - I even got issue with stock fw ....

**Comment by superdavex** [comment]
Here is the LCD http://www.winstar.com.tw/products_detail_ov.php?ProID=30  its a Winstar 1604a rev g

**Comment by luc-github** [comment]
ho you opened your printer ?
I found a spec related to it https://nettigo.pl/attachments/60, rom character looks is this one

so need to do init sequence according the spec

**Comment by luc-github** [comment]
I need to go out my wife is pushing, sorry - I will update this afternoon (your tomorrow)

**Comment by superdavex** [comment]
Yeah,  wasn't to hard to get to. Hehe, no worries same here.

**Comment by luc-github** [comment]
Sorry Just back home - will work on change tomorrow 
I was able to read spesc on my phone and sequence is not what we use today, and I am not sure all commands are identical to the #define ones in ui.cpp

actually reading a it, I realized I found similar sequence before (https://github.com/luc-github/Repetier-Firmware/issues/37#issuecomment-63425620) but never took time to test it .....

**Comment by garry572** [comment]
Ok buys days so wont have much time until the weekend. have had a quick look at the data sheet, not too far removed from the HD44870 but small changes can be significant

**Comment by garry572** [comment]
opps excuse my typos

**Comment by luc-github** [comment]
Yes will work on it today

**Comment by luc-github** [comment]
I make some violence to myself as you said it was not too difficult to check screen, I removed sides and front pane,l and actually you right, it was pretty easy.
 I unscreewed the LCD/key panel and no useful indication, LCD is soldered so not easy to check back side, I tried to unfold a little the LCD board and keypad board and .... I could see a sticker beetween boards : winstar and on board itself: 1604A rev G....

So looks like we have same LCD/Panel
![p_20150107_162335 jpg](https://cloud.githubusercontent.com/assets/8822552/5642826/1ea26f40-968b-11e4-96eb-cb016a0ffee2.jpg)
![p_20150107_162449 jpg](https://cloud.githubusercontent.com/assets/8822552/5642836/51efa6c4-968b-11e4-8c37-83d51ce6cb80.jpg)

I reviewed commands and sequence on specs but it will be more easy to test on my screen first

**Comment by garry572** [comment]
Hi Luc
Got the better of me, looking at the connection diagram the LCD panel had and 8 bit data connection, but all the firmware only used 4 bit mode. Ive added extra code to use the display in 8 bit mode, and works the same as the 4 bit (like it should) Now we have the choice of either mode. I will do some testing , but the only obvious change is the display seems slightly faster in response

**Comment by luc-github** [comment]
Hi,
I took the opposite direction : I use the 4bits functions/code we have, and just change the commands send to LCD using the 4bits ones

Testing also and yes start is definitly faster

**Comment by luc-github** [comment]
One thing, but you may be know for sure:
 to test properly you must flash FW - then  switch off printer and unplug USB/(power cord also ?) and wait few minutes to be sure no init information stay in LCD

And only after few minutes you can replug and power on to verify the change are good.

I got wrong result when just testing without discharge information of LCD, code looks good but actually after unplug and waiting the screen was white or not initialized

**Comment by garry572** [comment]
Yes i know :) the reason I wanted to try 8 bit mode, I have some older displays, that, although they are meant to work in 4 bit mode, aren't reliable. Also with 8 bit, each byte is written to the display only once, not split into 2 4 bit nibble and written twice. I changed the code, so in can be compiled with either 4 bit or 8 bit mode. I tested mine, maybe 20 times, turn off, wait  a few minutes, turn on. so far no failures its also hot here. 30C so i cant test if its reliable at low temperature

**Comment by superdavex** [comment]
Same here, I tried a couple sequences and it really didn't seem to make much difference.  The worst ones are the ones where the text garbled or had bad characters.  But with one close to the above sequence the only thing it would do is occasionally have a blank screen.  I tried dramatically increasing the timings, but, never tried reducing them.

**Comment by blackmambazaa** [comment]
Is it a specific set of events that result in the LCDs failing? Sorry, but
when I pushed the 0.92 firmware to my 1.0a I hadn't had the LCD failing.
This had been done over a week ago now and I have done some really long
prints as well.

On Wed, Jan 7, 2015 at 4:20 PM, superdavex <notifications@github.com> wrote:

> Same here, I tried a couple sequences and it really didn't seem to make
> much difference. The worst ones are the ones where the text garbled or had
> bad characters. But with one close to the above sequence the only thing it
> would do is occasionally have a blank screen. I tried dramatically
> increasing the timings, but, never tried reducing them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-69036056>
> .
>



-- 
Kind Regards

Tim Yagan

Tel: 44 7454 190922

Email: timyagan@gmail.com

**Comment by luc-github** [comment]
to be honest no real idea why same lcd has low failure on some system and high on others, but what is sure is after identifying the lcd model and then the correct command, it is definitly pure luck that current lcd initialization is working

one trigger to reproduce failure is cold boot and low temperature

**Comment by superdavex** [comment]
Once the display is initialized successfully it doesn't matter how long it is on, it works fine.  It's only from cold power on that occasionally it wail fail to initialize. An emergency stop fixes it every time.

**Comment by blackmambazaa** [comment]
Ah ok. I will keep an eye out for that on my printer.

On Wed, Jan 7, 2015 at 4:34 PM, Luc <notifications@github.com> wrote:

> to be honest no real idea why same lcd has low failure on some system and
> high on others, but what is sure is after identifying the lcd model and
> then the correct command, it is definitly pure luck that current lcd
> initialization is working
>
> one trigger to reproduce failure is cold boot and low temperature
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-69038334>
> .
>



-- 
Kind Regards

Tim Yagan

Tel: 44 7454 190922

Email: timyagan@gmail.com

**Comment by Doudou14** [comment]
I have this problem about once in five when I turn on my cold printer (Especially if it is not running for several days)
I think about a problem at the sending commands on the screen, something must cause a shift of the characters. (I have two black lines and the rest have realize characters)

**Comment by luc-github** [comment]
Edouard you can try the testing branch , it has the correct initialization for lcd, so normally no more faillure, finger crossed...

**Comment by superdavex** [comment]
Still seeing the same failure with the new changes.  Its not very often and easy to get back to normal with an emergency stop, but, every now and then there is a blank screen after boot up. This did not occur with the factory rom.

**Comment by luc-github** [comment]
did on my side also and getting worse actually

**Comment by luc-github** [comment]
So I put back current testing and also put back one timing in uiconfig.h 
change UI_DELAYPERCHAR	50 back to 320

**Comment by luc-github** [comment]
I was able to have 1 white screen so I try back to duplicate initialization sequence as first sugestion of Garry - then worse - really worse
right now the best is current testing code may be with UI_DELAYPERCHAR  320

I have changed also 
lcdWriteNibble(0x08|0x00); //Lines (0x08) and Font(0x00)
to
to lcdWriteNibble(0x08|0x04); //Lines (0x08) and Font(0x00)
as I am not sure what value is supposed to be for this one and no effect

so last test would be to switch to 8 bit communication to decrease communication issue due to 2 loops are necessary for one command 

Garry can you share what you did ? I can also do but as you said you already did - I could save my energy ^_^

Thanks

**Comment by luc-github** [comment]
I cannot reproduce white screen any more not sure it is due to 
UI_DELAYPERCHAR 320
lcdWriteNibble(0x08|0x04); //Lines (0x08) and Font(0x04)

or because less cold in room now

**Comment by superdavex** [comment]
Neither of those make much of difference for me.

**Comment by luc-github** [comment]
so 8bits communication may help to improve- wait for Garry code

**Comment by garry572** [comment]
Hi Luc will post it tonight, still at work :(

**Comment by luc-github** [comment]
No pressure ^_^, take your time

**Comment by luc-github** [comment]
looking at 
![image](https://cloud.githubusercontent.com/assets/8822552/5660380/7bc61238-975a-11e4-8b96-bc0e226b54ed.png)
Tc should be at least 1.2micro seconds (1200nanosec)
but in
```
void lcdWriteNibble(uint8_t value)
{
    WRITE(UI_DISPLAY_D4_PIN,value & 1);
    WRITE(UI_DISPLAY_D5_PIN,value & 2);
    WRITE(UI_DISPLAY_D6_PIN,value & 4);
    WRITE(UI_DISPLAY_D7_PIN,value & 8);
    WRITE(UI_DISPLAY_ENABLE_PIN, HIGH);// enable pulse must be >450ns
    DELAY1MICROSECOND;
    WRITE(UI_DISPLAY_ENABLE_PIN, LOW);
    DELAY1MICROSECOND;
}
```
I think with the 2 loops command due to 4bits, the timing is between 1.0~ 1.1, and not over 1.2 
so changing the last `DELAY1MICROSECOND;` by `DELAY2MICROSECOND;` should help when using 4bits communication, no ?

**Comment by garry572** [comment]
Hi Luc, Ive uploaded my 4/8 bit code to my gitbub account

**Comment by garry572** [comment]
opps github

**Comment by garry572** [comment]
Ive tried the code in both 4 and 8 bit mode with the display running as fast as the data sheet says had no failures in either mode

**Comment by luc-github** [comment]
Ok thanks a lot

**Comment by garry572** [comment]
It currently set to 8 bit mode and has my modified tables for the bed and thermistor

**Comment by garry572** [comment]
had a few issues uploading code to github all there now

**Comment by luc-github** [comment]
Hmm I am not sure to understand why you do this :
#define LCD_DISPLAYOFF      LCD_DISPLAYMODE &!(0x04)  /**<    Display off */
instead of 
#define LCD_DISPLAYOFF      LCD_DISPLAYMODE | (0x00)  /**<    Display off */
the command is only used to provide flags on not to erase flags, it give same result anyway as 
LCD_DISPLAYMODE is the "header" and is never in positions of the flags and won't be acording to spec - the flag won't be erased as it do not exists  - I am just curious

Another question, I see you provide extra bits like here:
 lcdCommand(LCD_CONFIGURATION | LCD_8BIT | LCD_2LINE | LCD_5X7);
LCD_CONFIGURATION  = 0X02 when command is * at this position so I guess "do not care" 
the result won't be affected as flag is ignored but any reason to provide it ?

Sorry for my questions, I am just wondering why

**Comment by garry572** [comment]
well basic logic you cant turn values of in a variable by using OR functions, OR can only turn bits ON
The AND function with an inverse mask will turn bits off ie 1 OR 0 still = 1 1 AND 0 =0 It you want to switch between function ie 4 and 8 bit mode you need to both set and clear flags.. Also it might be a don't care for 4 bit mode, but for 8 bit mode a lot of the lower bits have meaning

**Comment by garry572** [comment]
Also display off and display on both operate on the same it, so you need to be able to set and clear them. if you set the bit, you can never clear it by using OR

**Comment by garry572** [comment]
It may not matter , but Ive been caught before where a don't care of a data sheet really does matter

**Comment by garry572** [comment]
remember too the commands are now 8 bit so the upper bits need values

**Comment by garry572** [comment]
its also considered bad practice to set and clear whole bytes when you only need to change a bit

**Comment by garry572** [comment]
#define LCD_DISPLAYOFF LCD_DISPLAYMODE | (0x00) / will OR the value with 0x00 and will not change the state of the flag

**Comment by luc-github** [comment]
Yes I agree when to handle register, but in this case you handle commands that you already intialized by LCD_CONFIGURATION or LCD_DISPLAYMODE or LCD_ENTRYMODE that already set all bits but the command one to 0, so there are no flags to set off and the & won't be applied to other command addeds by the |
if you want to flag off the bit you should not use the LCD_CONFIGURATION in the define because it limits its action to the define not to the command
#define LCD_CONFIGURATION       0x20                /**< Set function */
#define LCD_5X7         LCD_CONFIGURATION &!(0x04)    /**<    5 X 7 dots */

because 
in lcdCommand(LCD_CONFIGURATION| LCD_8BIT | LCD_2LINE | LCD_5X7);
the & has no action on others flags and even to this LCD_CONFIGURATION only on the one in defined

It is not big deal anyway - thanks

**Comment by garry572** [comment]
well yes and no the value of LCD_CONFIGURATION has a value of 0x20 and the bit mode, line mode or entry mode don't set or clear more than their own bits what the state of LCD_CONFIGURATION is preserved. Beside there no sense ORing a number with 0 you just get the same result. so why do it.

**Comment by luc-github** [comment]
Just for readability actually - I agree useless

**Comment by garry572** [comment]
yes :)

**Comment by luc-github** [comment]
my printer is still busy printing and did not finished the 8bits modification
will work on it and test it tomorrow
I have updated the 4bits functions to match the specs on testing branch, so far no white screen but this is not a definitive result :D 
On top of adding the 8ibt support, I will update the 4bits sequence in same way as Garry did, because it is more elegant to use the lcdCommand than the lcdWriteNibble, and more easy to read, 

I will also modify some #define for readibility like font flag (5x8 or 5x11)

**Comment by Doudou14** [comment]
Did you watch the three file I pushed?

**Comment by luc-github** [comment]
Edouard, actually we already integrated/tested these changes in testing branch before you pushed https://github.com/luc-github/Repetier-Firmware-0.92/commit/e6784ade5a12397d65aeed007b03357da519aa1b

**Comment by luc-github** [comment]
Testing branch is updated with 8bits support based on Garry code
I have used all timing from Garry but the one between commands, I have increased it from 60 to 100
because I got one scramble screen, once in 8bits and once in 4 bits, so far no more until now....

please have a test and feedback - hope it will improve situation

**Comment by luc-github** [comment]
this morning cold start - and several restart  - no white neither scramble screen - tried several times (~20) keep testing

**Comment by garry572** [comment]
had one LCD initialize failure today, cold start temperature was 23C Increased delay between commands to 150

**Comment by luc-github** [comment]
Ho ~ using 4bits or 8bits ?

**Comment by garry572** [comment]
Using 8 bit increased the delays a little, no more failure :)

**Comment by luc-github** [comment]
Ok I will increase on testing code so

**Comment by garry572** [comment]
also testing to insure that the hardware produces the delays we expect, between 1usec and 6000 usec, close enough not to contribute to the errors

**Comment by luc-github** [comment]
Ok great - thanks
150 is in now https://github.com/luc-github/Repetier-Firmware-0.92/commit/0d491e15da697a2b11eca0850af23d8e79c57588

**Comment by garry572** [comment]
You only using 4 bit ? the pushed changed are only on the 4 bit

**Comment by luc-github** [comment]
no both - code is common now

**Comment by luc-github** [comment]
only header and flags for bits size are different
so I put compilation condition on header and flag value

so no need to duplicate sequence and more easy for update like timing

**Comment by garry572** [comment]
Glad to see you tidied up my code :)

**Comment by luc-github** [comment]
you did all the job but I was lazy to copy all :p

**Comment by garry572** [comment]
Yeah well I was too lazy to tidy it up too, I had the code separated to see if there was anything different between 4 and 8 bit :)

**Comment by superdavex** [comment]
Looks like the 8 bit change did the trick. I cannot get mine to fail now.  Tried 20 to 30 times.  Even several quick off and ons that would always fail before.  Not a single failure.

**Comment by luc-github** [comment]
cool, what timing do you use 100 or 150？

**Comment by superdavex** [comment]
I pulled down the test branch full again so 150.

**Comment by luc-github** [comment]
excellent then, I suggest we test and if still ok on Monday, I will push to master

**Comment by superdavex** [comment]
agreed, sounds good.  Nice job guys!

**Comment by garry572** [comment]
Has a failure on my display this morning, Cold morning (22C) start up was OK, When I tried later, if wouldn't d start after 4 or 5 tries. Then it came good and wont fail again.I beginning to think as well as cold start up issues. the displays are generally of poor/cheap quality

**Comment by luc-github** [comment]
no issue on my side so far, i tried every morning around 20 times, during the day several times, and late night also, I am out now for all day, will try tonight, and after some print

**Comment by garry572** [comment]
Think I found another potential cause of LCD failure. As you know the IO pins are programmable, so that to use the LCD, the data pins and lcd enable as rs select are set for output mode. By default at power up, until programmed they default to input mode. This means they are either at an indeterminate state, or pulled up to supply. I found that if i set all the data pit on the LCD port to 0, before the enable pin is set, first to an output, then to a low state, the display never initializes. Tried over 50 times and each time the display was never initialized correctly. I then tried setting all the LCD pins to 1, and it never failed once in 10 tries. Making them 0 and it always fails. My suspicion is that as the enable pin is set to output and set low, the led interprets the random data as a command.

**Comment by garry572** [comment]
Ive moves all the logic in the initialization , so that the LCD enable is the last set of the pins, so that known data is on all the LCD data pins as well as the RS pin Doing this I can go from 100% failure to no failures

**Comment by luc-github** [comment]
Ok, I will check when home tonight

**Comment by garry572** [comment]
Found I needed to make the command for the first change of state of the LCD enable to be something thats not used in the init sequence, any of the commands that read from the LCD RAM are OK as they don't interfere with the LCD initialization.

**Comment by luc-github** [comment]
you mean enable pin do a high and low without any command then start init commands ？ like kind of warmup for lcd

can you share the modified part ?

**Comment by garry572** [comment]
yes that's I mean

   SET_OUTPUT(UI_DISPLAY_D5_PIN);
    SET_OUTPUT(UI_DISPLAY_D6_PIN);
    SET_OUTPUT(UI_DISPLAY_D7_PIN);
    SET_OUTPUT(UI_DISPLAY_RS_PIN);
    
    WRITE(UI_DISPLAY_D0_PIN, HIGH);
    WRITE(UI_DISPLAY_D1_PIN, HIGH);
    WRITE(UI_DISPLAY_D2_PIN, HIGH);
    WRITE(UI_DISPLAY_D3_PIN, HIGH);
    WRITE(UI_DISPLAY_D4_PIN, HIGH);
    WRITE(UI_DISPLAY_D5_PIN, HIGH);
    WRITE(UI_DISPLAY_D6_PIN, HIGH);
    WRITE(UI_DISPLAY_D7_PIN, HIGH);
    
    
#if UI_DISPLAY_RW_PIN>-1
    SET_OUTPUT(UI_DISPLAY_RW_PIN);
#endif
    
    WRITE(UI_DISPLAY_RS_PIN, LOW);
    HAL::delayMicroseconds(5);

    SET_OUTPUT(UI_DISPLAY_ENABLE_PIN);
    HAL::delayMicroseconds(5);
    WRITE(UI_DISPLAY_ENABLE_PIN, LOW);
    //initialization sequence for 4bits/8bits of Winstar 1604A Screen
    //16 rows, 4 lines
     HAL::delayMilliseconds(10); // Just to be safe, not sure how long command takes
     HAL::delayMicroseconds(5);

#if  UI_DISPLAY_TYPE == DISPLAY_4BIT
    lcdWriteNibble(0x03);//Init Function Set for 4bits
    HAL::delayMicroseconds(150); //more than 39micro seconds
#endif
  
This shows where  I made the changes, just after the LCD pins are defined as output I set them to a known state

The just as the LCD pin is defined, then enabled I put a delay in to make sure its completes

I had no failures at all until the temp got to 25C then magically the LCD acted silly.

Seems if I set the LCD output values to anything that's the same as an initialize command. I can go from 100% working to 100% not working.  Just shows that if its repeatable it easier to find a cure

**Comment by garry572** [comment]
I will push the changes tomorrow, too late to think now :)

**Comment by garry572** [comment]
no i don't do and LCD enable pin high , I suspect it's already high on power up, as programmed its an input, if it was already low then the commands do nothing

**Comment by luc-github** [comment]
I will try this tomorrow - just back and too late now - good catch

**Comment by superdavex** [comment]
17c in here this morning, 25 times off/on.  No failures.

**Comment by blackmambazaa** [comment]
I also haven't had any issues thus far with similar sort of test values in
my work environment without any failure and this is with the 0.92 firmware
that was released a while back.

On Sun, Jan 11, 2015 at 4:26 PM, superdavex <notifications@github.com>
wrote:

> 17c in here this morning, 25 times off/on. No failures.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-69498065>
> .
>



-- 
Kind Regards

Tim Yagan

Tel: 44 7454 190922

Email: timyagan@gmail.com

**Comment by luc-github** [comment]
That the problem with random issue ^_^ some people have and some don't have so hard to status
I think we are close took.
I do not have aby failure neither, but as garry meet one - problem can popup anytime anywhere if it is due to unknowstate of pins - but rate can be very low so not seen by everyone quickly

**Comment by garry572** [comment]
Your right Luc, it possible that the problem may never exist, but  for me, after 3 days of no failures, it happend to me yesterday. I can force a 100% failure by writing certain values on the LCD pins before initialization. It possible that no one else will have this issue. but for me, based on the work Ive done in the past, even 1 in 1000  failures is too many, so I decided to find the root cause of my issue.

**Comment by luc-github** [comment]
I will test the change today

**Comment by luc-github** [comment]
I made change in testing (https://github.com/luc-github/Repetier-Firmware-0.92/commit/a7c779d7daa81c048ead6d05af834ee0d40af630) and testing now

**Comment by garry572** [comment]
Hi Luc , any updates

**Comment by luc-github** [comment]
Still no white screen / no scramble so far ^_^

**Comment by garry572** [comment]
same here , no failures, no white screen, no tempertaure related issues :)

**Comment by luc-github** [comment]
as this change is no affecting timing I guess should be "working or not".
So I suggest to wait superdavex feedback on this one, and if no issue neither - I will push in master branch and we should be good

**Comment by garry572** [comment]
sounds good

**Comment by superdavex** [comment]
Tried 30+ times at 20c, no failures.  I will try again in the morning when its cooler, but, temperature was never an issue for me so I don't expect any issues.

**Comment by luc-github** [comment]
Ok thanks a lot - waiting for your confirmation then

**Comment by garry572** [comment]
OK another day, no   issues

**Comment by garry572** [comment]
The only change I can suggest is the delay after the LCD Enable pin goes low should be at least 40 usec, its currently set to 5usec which is too short for the commands I made mine 10msec to be safe, as the example I posted above

**Comment by luc-github** [comment]
Yes saw it and forgot to ask,
Why 5microsec is too short ? all spec is about 39 nanosec 
because 10milsec is very long

**Comment by garry572** [comment]
no its not 39 nsec is 39 usec

**Comment by garry572** [comment]
actually the command I used takes 43usec

**Comment by garry572** [comment]
I used 10 msec since all other delays are shortened and you only initialize once during power up, better to be safe than sorry

**Comment by luc-github** [comment]
Ho I see - the write data to ram = 43micro sec 
I will update this timing to 10ms then

**Comment by luc-github** [comment]
Implemented (https://github.com/luc-github/Repetier-Firmware-0.92/commit/69c3cd3d042375e5190f582d3f763fa7f476534c) so far no obvious white screen neither scramble

**Comment by garry572** [comment]
same here lots of re starts no errors/white screen

**Comment by luc-github** [comment]
Got some failure reproductible 100%, can be a corner case, but may confirm the possibility of issue for unknown/wrong pin state

I was working to check an issue so I use bgm test sckech because light and easy to modify, then I flashed back to repetier (testing branch) and got scramble screen, one restart and screen was ok
I did it a second time (flash test sketch then flash repetier (testing)) and issue again 
I did it a third time (flash test sketch then flash repetier (master)) and no issue
I did it a fourth time (flash test sketch then flash repetier(testing)) and issue again 
I did it a 5th time (flash test sketch then flash repetier (master)) and no issue
I did it a 6th time (flash test sketch then flash repetier (master)) and no issue
I did it a 7th time (flash test sketch then flash repetier (master)) and no issue
I did it a 8th time (flash test sketch then flash repetier(testing)) and issue again 
I did it a 9th time (flash test sketch then flash repetier(testing)) and issue again 
I did it a 10th time (flash test sketch then flash repetier(testing)) and issue again 

flashing second time repetier(testing) when issue do not solve the problem, need restart or emergency stop

This is may due to fact there are no restart between flash but this also confirm the risk for unknown state

**Comment by superdavex** [comment]
Tried again this morning at 17c, no issues.  I had seen failures after flashing the last two patches.  It was only right after flashing with no restart.  I just attributed it to a bad state.  I have had over 70 to 100 restarts, some rapid fire and no failures.

**Comment by luc-github** [comment]
OK seems linked to flash first LCD in 4bits then in 8bits
if I flash testing branch with 4bits enabled after bgm sketch, no issue, and I confirmed flashing after same testing branch in 8bits, and issue happen
So corner case - should not be a problem

So based on superdavex feedback I think we can merge to master and add note that a restart may be necessary after flashing

**Comment by blackmambazaa** [comment]
You guys rock! I wanted to write some test cases for this, but I had no
time whatsoever :(

On Mon, Jan 12, 2015 at 3:58 PM, Luc <notifications@github.com> wrote:

> OK seems linked to flash first LCD in 4bits then in 8bits
> if I flash testing branch with 4bits enabled after bgm sketch, no issue,
> and I confirmed flashing after same testing branch in 8bits, and issue
> happen
> So corner case - should not be a problem
>
> So based on superdavex feedback I think we can merge to master and add
> note that a restart may be necessary after flashing
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-69581474>
> .
>



-- 
Kind Regards

Tim Yagan

Tel: 44 7454 190922

Email: timyagan@gmail.com

**Comment by luc-github** [comment]
Ha test cases would be great but yes no time ^_^

**Comment by garry572** [comment]
Still no issues on my side,  For my 1.0 restart is needed after flashing as mine will not directly after flash

**Comment by garry572** [comment]
Part of the reason for LCD initializion failures can be due the LCD cannot change some modes ie 4/8 bit mode can only be changed by power up commands not by commands ater. A restart is recommneded to ensure the LCD is always reset correctly

**Comment by luc-github** [comment]
I will push changes to master branch today, it is nice to work with you guys - really appreciate
Thanks a lot to make this fw better ^_^

**Comment by luc-github** [comment]
pushed (https://github.com/luc-github/Repetier-Firmware-0.92/commit/195076d1621c9e6d4b95288737e7284fdc462908)
Thanks again everyone

**Comment by jnadke** [comment]
LCD Initialization still fails on the latest "master".

**Comment by luc-github** [comment]
I just revert the last change in testing can you have a try ?

**Comment by jnadke** [comment]
I can get it to reliably fail just by turning the machine on and off.  I am uploading the latest changes from testing now.

**Comment by luc-github** [comment]
Ok thanks

**Comment by jnadke** [comment]
Latest change does not fix the issue.

I even double checked ui.cpp for the commented line of code cross referenced to GIT online to sanity check I was in the right branch.

**Comment by luc-github** [comment]
I just revert more - I removed the pin initialization to see if could be the root cause

**Comment by jnadke** [comment]
That does not work either.

**Comment by luc-github** [comment]
sorry to ask - do you restart after flashing ?

**Comment by jnadke** [comment]
Yes, I've power cycled many times (but did not issue the M502/500 command, I assume that's not relevant).

**Comment by blackmambazaa** [comment]
Did you save the setting on your printer via the on screen menu?

On Wed, Jan 21, 2015 at 9:29 AM, jnadke <notifications@github.com> wrote:

> Yes, I've power cycled many times (but did not issue the M502/500 command,
> I assume that's not relevant).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-70800947>
> .
>



-- 
Kind Regards

Tim Yagan

Tel: 44 7454 190922

Email: timyagan@gmail.com

**Comment by jnadke** [comment]
I wish I could, the on screen menu works maybe 1 in 10 tries (I just tried 15 times and got it to work twice).

**Comment by luc-github** [comment]
Yes M502/M500 is not necessary for LCD init

**Comment by luc-github** [comment]
I just update testing again - putting back a typo issue that may affect initialization - as after that I am dry what to test as cannot reproduce issue.....

**Comment by garry572** [comment]
Ive been using mine for over a week with temperature between 20C and 30C with no failures, so not sure what the issue is, I still have the double initializelcd() so maybe that's a clue

**Comment by jnadke** [comment]
I'll try your last test, otherwise it's time for me to sleep.

I can look at it tomorrow.  I just pulled up the data sheet (you said it was a HD44780U) as well as a small guide on the proper initialization.

**Comment by jnadke** [comment]
I just tried your latest update (3c966ae) and it did not work.

**Comment by luc-github** [comment]
Nope it is a winstar 1604A but initial statement was it is a HD44780 clone - but was wrong
tomorrow you can try the 0.91 it use HD44780 way

**Comment by luc-github** [comment]
Yes will add double init in testing branch for you test tomorrow also -

**Comment by luc-github** [comment]
Revert to master code and duplicate initializeLCD implemented in testing branch 
Damned LCD - never ending story....

**Comment by jnadke** [comment]
So, if I power off, wait 2-3 seconds, and power back on, usually it doesn't work.  But if I power off/on very rapidly (< 0.5 sec off/on), it works fine.

I assume that's less than the discharge time of any capacitors, could dirty power be causing the initialization commands to be not accepted?  The system would be taxing the supply I'd imagine on startup.  I see in the data sheet it has a very long startup period (15ms).

Either way, I'll look at it more tomorrow.  Thank you.

**Comment by jnadke** [comment]
I should clarify the rapid restart trick works even if the LCD previously was not working.

So either some preliminary initialization isn't being obeyed by the LCD or its a vcc rise issue where possibly at boot up the lcd is still initializing and rejecting commands.

Of course this is easily testible by delaying the initialization.  If the issue is rejecting commands, then 2x init would fix it (though the proper fox would be to check the read busy flag before sending a new command).

**Comment by garry572** [comment]
To read the state of the busy pin, you need to put the LCD in read mode. This requires the R/W pin to be connected to the CPU. Since the IO list I have doesn't show the R/W pin I assume its always low and the LCD is always in the write mode.

**Comment by ogremustcrush** [comment]
Can confirm same behavior as jnadke on latest commits, no initialization at all unless you do rapid power cycle. 1.0a model.

**Comment by garry572** [comment]
The issue with these LCD's is they need a defined rise time of the power supply to reset their internal operations. if this isnt correct, they may not reset correctly internally. If this happens, no amount of changing state of pins will help. larger / other displays have a dedicated reset pin that removes the power supply start up requirments. The combination of LCD,s and power supplys makes resolving the issue very difficult.

**Comment by jnadke** [comment]
I assume the rise time isn't an issue or the original firmware would not function at all.  The issue is definitely programming.

The main thing is ensuring that something in the hardware is not rendering the assumptions behind the programming invalid.

**Comment by luc-github** [comment]
So solution is to find the RW pin and implemented it ?

testing branch has the 2x init anyone having the issue tested it ?
if it is rising power issue 
will modifying 
```// SEE PAGE 45/46 FOR INITIALIZATION SPECIFICATION!
    
    // according to HD44780 datasheet, we need at least 40ms after power rises above 2.7V

    // before sending commands. Arduino can turn on way before 4.5V.

    // is this delay long enough for all cases??

    HAL::delayMilliseconds(500);```
to 
``` HAL::delayMilliseconds(2000);```

will help ?
since emergency stop or rapid restart works I guess the timing for initialization but the first one are ok

**Comment by garry572** [comment]
Interesting, the Due has a brown out detector, so rise time isnt a concern for it. The LCD isnt quite as smart.

**Comment by garry572** [comment]
One interesting thing I did find, when I was testing the accuracy of delays on my Arduino Due, I loaded the 0.92 code into it, and put a toggle an I/O pin in the main code. After about 10 seconds, no matter what pin I used the pin stopped toggling. When I stopped the main prgramiing being called this stopped. All I could assume, is the even unallocated pins are somehow modified.. This will need more investigation.

**Comment by jnadke** [comment]
I will test your 2x init code in 2-3 hours.

**Comment by jnadke** [comment]
The 2x init code makes the LCD function maybe 50% of bootups.

**Comment by luc-github** [comment]
Ok I have increased the timing to let power rise to 2 seconds and left the 2x init 
boot should be slow but at least it should allow to check if issue is due to power rise

next step would be to find the RW pin and implement it - screen support it, just need to find where this pin is connected

**Comment by luc-github** [comment]
modification done in testing branch https://github.com/luc-github/Repetier-Firmware-0.92/commit/5300b3b52347088388aac877445e3511c806bf54

**Comment by jnadke** [comment]
I still get failures, it's probably not power-up related.  There must be some issues with it not liking some of the commands some of the time.

**Comment by jnadke** [comment]
So, I've been referencing this datasheet:
https://www.spezial.com/doc/winsta/1604a-tfh-jt+_rev_01.pdf

Which outlines the proper initialization on page 18 (8-bit mode).

**Comment by luc-github** [comment]
yes we used this one already, check previous messages : https://github.com/luc-github/Repetier-Firmware-0.92/issues/10#issuecomment-68820979

**Comment by luc-github** [comment]
the only difference is the RW pin is not used as not defined so BF is not tested but a timing is set instead

**Comment by luc-github** [comment]
I can try to figure out which pin is it,  but I need to open printer and test LCD pins continuity, etc... and now I am printing - with at least 6h to go - so if someone can find it more fast it is ok 

So far I have 0 failure on my Duo since we implemented the 8bits sequence from winstar specs, and testing on 1.0, 1.0A and 2.0 were good also - I am still wondering what is the difference among the 1.0A units that make some of them LCD Init failing and make others successful

**Comment by jnadke** [comment]
I haven't had time to open my printer, but I looked at the lcdWriteByte function and slowed down the operations a bit.  I also moved the Enable pin to after the Data pins get set to ensure they are at their proper values.

The incidence of failure has gone down significantly.

```
    WRITE(UI_DISPLAY_RS_PIN, rs);
    HAL::delayMicroseconds(5);
    WRITE(UI_DISPLAY_D0_PIN, c & 0x01);
    WRITE(UI_DISPLAY_D1_PIN, c & 0x02);
    WRITE(UI_DISPLAY_D2_PIN, c & 0x04);
    WRITE(UI_DISPLAY_D3_PIN, c & 0x08);
    WRITE(UI_DISPLAY_D4_PIN, c & 0x10);
    WRITE(UI_DISPLAY_D5_PIN, c & 0x20);
    WRITE(UI_DISPLAY_D6_PIN, c & 0x40);
    WRITE(UI_DISPLAY_D7_PIN, c & 0x80);
    HAL::delayMicroseconds(5);
    WRITE(UI_DISPLAY_ENABLE_PIN, HIGH);   // enable pulse must be >450ns
    HAL::delayMicroseconds(10);
    WRITE(UI_DISPLAY_ENABLE_PIN, LOW);   //tc must be >1200ns
    HAL::delayMicroseconds(5);
```

I also did revert change 5300b3b52347088388aac877445e3511c806bf54 but kept the 2x init.


It might still be productive to hunt down the RW pin (if it's even connected), but I'm happy for now.

**Comment by luc-github** [comment]
Ok Great thanks a lot 
I will test your modification when my printer is available to see if ot does not affect in other side the no 1.0A

**Comment by luc-github** [comment]
I was not able to find the RW pin, I was juste able to test RW pin is not connected to ground.

I have implemented the modification of jnadke : https://github.com/luc-github/Repetier-Firmware-0.92/commit/ecdd250a31ef3d6aa9c07a537e704c61e9ca96f6
=>one additional `HAL::delayMicroseconds(5);` before enable
=>one extended from 5 to 10 `HAL::delayMicroseconds(10);` after enable
but without double initialization and without big timing for raise power (so like in master) - to see if they are mandatory or not 

I have tested on my 2.0SF and at least it did not bring me any failure
Please have a check and feedback - Thanks

**Comment by oscahie** [comment]
I'm not sure if the issue is now only limited to the 'A' models, but my 1.0 (non-A) has not had a single LCD initialization issue since the fix was applied to master. It's working great.

**Comment by luc-github** [comment]
Looks like issue is on some 1.0A not all - yes I have no issue neither, so it is hard to fix it but if jnadke cover all it would be great

**Comment by jnadke** [comment]
So, I did some testing, and I'm 99% sure PIN 100 is the RW pin.


1. There are some ICs under the LCD that I believe are pullup/pulldown resistors packs or amplifiers, for amplifying the weak signals traveling across the ribbon cable.  I couldn't read the part number.

2.  One side is exposed, which are signals going to the ribbon cable, the other side goes to the LCD (which is difficult to get at).

3.  I looked at the ATMEL processor and verified PIN 133 (RS) went to PIN 20 in the ribbon cable.  Pin 19 on the ribbon cable went to PIN 100.  I verified PIN 17-10 went to the data pins.  PIN 9 goes to PIN 39 on the ATMEL (isolation ground?) and the others are the button pins.

4.  Anyhow, I verified on the LCD module that RIBBON CABLE Pin 20, 19, 18, etc all go to different pins on the amplifier ICs.  Specifically 20 and 19 went to two pins next to each other.  And 18 to another pin.

5.  Finally, I took a piece of bare wire and wrapped it around my multimeter to "rake" the other side of the ICs under the LCD.  I started with RS and sanity checked, my multimeter gave continuity at a pin approximately directly across from what I found in step 4.  I then verified the ENABLE pin, same thing.  Lastly, I tried the RW pin.  BEEP.  It's hooked to the amplifier ICs.


We probably don't even need to fully implement RW for BUSY FLAG.  If we just set it at OUTPUT-LOW and lock it there.

I suspect RW is floating right now (uninitialized pins become pullup-resistor inputs) leading to some commands being rejected when its interpreted as 1.  Though it's puzzling that character display works fine.

**Comment by jnadke** [comment]
Sure enough, after disassembling everything I reloaded MASTER (without all the hacks) onto my printer and the LCD initialized perfectly.

Anyway, I threw some code into my personal branch that did:
```
    HAL::delayMilliseconds(500);
    SET_OUTPUT(UI_DISPLAY_D4_PIN);
    SET_OUTPUT(UI_DISPLAY_D5_PIN);
    SET_OUTPUT(UI_DISPLAY_D6_PIN);
    SET_OUTPUT(UI_DISPLAY_D7_PIN);
    SET_OUTPUT(UI_DISPLAY_RS_PIN);
	SET_OUTPUT(45);
```

And:
```
    // Now we pull both RS and R/W low to begin commands
    WRITE(UI_DISPLAY_RS_PIN, LOW);
	WRITE(45, LOW);
    HAL::delayMicroseconds(5);
```

My printer did not start on fire.

That's about all I can verify.  And the LCD works 100%, but like I said, after I reassembled everything I loaded virgin MASTER and it worked the 4-5 times I turned it off/on then too.

**Comment by luc-github** [comment]
Wow, impressive
Sorry to ask but I want to be sure 
so put due pin 100 / arduino 45 to low solve the issue ?
or actually you cannot reproduce the issue after dissembling/reassembling the printer using clean master branch ?

**Comment by jnadke** [comment]
Both.  I cannot reproduce it any longer before or after.

But Arduino 45 low doesn't break my printer.

**Comment by luc-github** [comment]
Understood - I will set your code in testing branch and see if someone having issue can test also to see if it is solving problem

Thanks a lot for your investigation ^_^

**Comment by jnadke** [comment]
Take out 2x init from the testing branch and undo my last changes that added delays to lcdWriteByte as well.

Those were just hacks and should not be necessary.

**Comment by luc-github** [comment]
Yes will use master as base

And I also think use 45 as RW may be tricky,I set RW pin to 45 and system display ok at first sight.
but ..  every second there is a refresh for temperatures and some characteres are shifted right or left
So need tuning 

I will test the 45 low now

**Comment by jnadke** [comment]
Yeah, I avoided that define because I haven't examined that code to see if it is correct.

**Comment by luc-github** [comment]
Ok my printer is not on fire neither and no visible side effect 
I have implemented your suggestion https://github.com/luc-github/Repetier-Firmware-0.92/commit/01d34a5d63d6106089842c595061936553775ec3
Let see if someone having issue can double confirm it solve it

**Comment by garry572** [comment]
Ok done some investigation. if you set  #define UI_DISPLAY_RW_PIN to anything else but  -1 the RW pin is enabled, also the busy check is set. This has not been confirmed as working The display clears but the characters are moving sideways . I used a dummy variable name to set pin 45 to output,  if i set it HIGH the screen never initializes if I set it LOW the screen seems to work ok. Seems that pin 45 is the RW pin of the LCD

**Comment by jnadke** [comment]
> set pin 45 to output, if i set it HIGH the screen never initializes

This is an excellent test.  I did not have a chance to test that.  Sounds like it's confirmed.

**Comment by jnadke** [comment]
As for UI_DISPLAY_RW_PIN, I'd like to have it implemented (currently we're delaying for 0.5 ms in lcdWriteByte which for 34 characters is 17ms.... which is 17ms the printer is not reading the sensors), but I really don't know what those IC chips are under the LCD.  I don't know if:
1.  They relay signals in the opposite direction
2.  Their specs.

To get UI_DISPLAY_RW_PIN / busy flag to work on 45 we probably need to activate the ATMEL internal pull-up/pull-down resistors.  Just setting it to INPUT will probably produce undefined behavior -- floating.

I would not want to do that without knowing what those chips do.  Chance of damage are very slim though, there's plenty of resistance in the path.  But I'm not that enterprising.  You'd essentially add to lcdWriteByte after SET_INPUT(UI_DISPLAY_D7_PIN):
PULLUP(UI_DISPLAY_D7_PIN, HIGH) -- or LOW, not sure which.


Beyond that, I'd be more in favor of moving back to 4BIT operation.  It's more efficient.  Less time writing characters to the display and more time checking the stepper motors so they don't travel too far.

**Comment by jnadke** [comment]
If we stick with 8BIT, I don't know what UI_DELAYPERCHAR currently is, but we should make sure it is at 40 (microseconds).

I was assuming it was at 320 microseconds above.   It doesn't need to be any greater than 40 microseconds.

**Comment by garry572** [comment]
The LCD is 8 bit internal, its only it data bus that can operate in 4 bit or 8 bit mode, writing in 4 bit modes means you need to write twice to send 8 bits to the display. If you looking to save time writing to the display, the 8 bit mode is more efficient of processor time

**Comment by jnadke** [comment]
My bad, I haven't been looking at that part of the datasheet.  I do see that 4 bit mode just lets you write 2 sets of 4 bit values.

So the only changes required:
-setting UI_DELAYPERCHAR to 40
-setting the other delaymicrosecond to 1 in lcdWriteByte (none of the data latches are more than 1000ns and the INIT/CLEAR functions are already protected by delays in initializeLCD()).
-create a new define for the RW Pin / document it and update the code to permanently latch it to LOW as above

**Comment by luc-github** [comment]
Yes 8 bits is faster as need less actions
About timing adjustment we tried to short ones and this did not worked well on some units
You can see history, we increased then actually, we tried UI_DELAYPERCHAR  40 then put back to 320 but this was before to get the RW set to low, so it worth to try again
And need to test on all generations.

I suggest to get feedback first with current timing/modification then merge to Master 
then generate a new revision in test branch with tighter timing to get feedback
It will also be easier to get comparaison like this

**Comment by jnadke** [comment]
Well, I forked the repository and threw in the changes for the lower delays.

All is working fine on my printer.

**Comment by luc-github** [comment]
Excellent ! sounds good

**Comment by garry572** [comment]
I one small issue the LCD always resets on power up, but an emergency stop makes the display blank. Putting back the double initializelcd fixes the issue

**Comment by luc-github** [comment]
you use the 3 init lines ？
I have tested and was ok
did you used the testing branch or you did changes manually ?

**Comment by garry572** [comment]
no i just used the double init,

**Comment by garry572** [comment]
Just tried your 3 init lines, they work well too, so I will leave them so it aligns with testing branch

**Comment by luc-github** [comment]
Ok excellent 1.0/2.0/1.0A ok - we still need feedback from another 1.0A owner and we are good

**Comment by garry572** [comment]
The set low for pin 45 wasn't in the testing branch

**Comment by luc-github** [comment]
you do not see it ? https://github.com/luc-github/Repetier-Firmware-0.92/commit/01d34a5d63d6106089842c595061936553775ec3

**Comment by garry572** [comment]
yes sorry missed seeing it all ok now

**Comment by luc-github** [comment]
No problem - I was worry my github has problem to sync ^_^

**Comment by disneysw** [comment]
First I have only just started playing with the printer and firmware and I have not read all the threads that are out there.

However I just installed the current Top of Tree (Version 1.15-01-02.1.1) and am having major issues with the LCD initialising after power on.

Prior to this I was using the V1 RC which only had the occasional initialisation problem.
Mine is a V1 board (with unpopulated J1 reset) dated 2013-12-26.

From reading this thread my understanding is everyone thought this issue was resolved on this version of the board?

I will do a bit of investigation myself after reviewing the LCD datasheet but would anyone like to make suggestions on what to try?

BTW, Interestingly the LCD always seems to work immediately after a reflash of the firmware (with power cycle) but fails 90% of the time on subsequent power cycles.

**Comment by luc-github** [comment]
Hi can you try the testing branch ? this one is suposed to solve the issue but need feedback

**Comment by disneysw** [comment]
Thanks, 

initial tests look like that does fix the problem!

**Comment by luc-github** [comment]
Thanks - please feel free to report any failure 
We are waiting several confirmations before merging to master as issue seems not affecting all units

**Comment by jnadke** [comment]
We should probably merge this into master sooner rather than later.  It's been confirmed to work on your DUO, 2 1.0's and my own 1.0A so far.

When I first flashed 0.92 on my 1.0A before fixing this issue, I seriously thought I'd broken my printer (particularly since the EEPROM light defaults to 0).

Someone less technically oriented might not be so calm-headed.

**Comment by mwilliams** [comment]
I'm awaiting merge of this to master before flashing my 1.0A.  Mine's dead in the water, rejecting every cartridge I throw at it and XYZ's support is terrible.  Time to mod!  But this issue has been the only thing shying me away...  Reports look good though.

**Comment by blackmambazaa** [comment]
Why don't you flash it to 0.91. Forget about the standard FW. I flashed to an early 0.92 master and haven't had much issues. Best thing that I could have done. My prints are much better now.

**Comment by luc-github** [comment]
Well before we tested on 1.0, 2.0 and 1.0A and no issue, so risk is still there without more confirmation.
Anyway as no one reported issues  (but 5 days is short) and seems not affecting badly the working units, I have moved the fix to master - hoping more feedback : https://github.com/luc-github/Repetier-Firmware-0.92/commit/bdc167473ba8003ac63b1bb9b789c4f5000ec873

**Comment by pmbdk** [comment]
Hey Luc, I havn't been able to use the LCD for the last two weeks since I updated to the latest version. But I just downloaded and installed Testing and it works perfectly. This is on the J37 version. But I can see that you put the changes into master, so I'll just give that a try as well.

**Comment by luc-github** [comment]
Thanks a lot for your feedback

**Comment by jnadke** [comment]
For what it's worth, I've been running UI_DELAYPERCHAR at 40 on my printer ever since this was fixed and haven't had any issues.

I don't know if it necessarily helps any for processor loading / print quality, but it's the correct setting for the longest delay the LCD requires.

**Comment by luc-github** [comment]
Ok - will update code later

**Comment by dlbyers** [comment]
I have a Duo J37 that was exhibiting the LCD problem.  I just updated to your new master and can confirm that the problem is resolved for me.

**Comment by luc-github** [comment]
Thanks a lot four your feedback

**Comment by luc-github** [comment]
master updated for UI_DELAYPERCHAR 40 : https://github.com/luc-github/Repetier-Firmware-0.92/commit/6d42a7aa9cd89cf29c7635cfed6f9bfa88e0896d

**Comment by luc-github** [comment]
seems 1.0 has scramble screen - so I have reverted back the last change of  UI_DELAYPERCHAR 40 to see if it is the root cause

**Comment by garry572** [comment]
I had the same trouble when I made  UI_DELAYPERCHAR 40 I had scrambled LCD characters, so  I left it at 320

**Comment by luc-github** [comment]
Ok so it confirm the revert

**Comment by jnadke** [comment]
Its unfortunate that 40ms causes issues for some.

I had a 1.0a and now a 1.0 and neither have issues with 40ms.

But leave it at 320.  I'll probably experiment with getting RW working.

**Comment by luc-github** [comment]
That is the difficulty of having several models - always have little difference
even small change can have big impact if not well tested
I am big follower of Murfy Law :)

**Comment by mwilliams** [comment]
Loaded master on my 1.0A yesterday, it's been running flawlessly. Bravo on
this...

**Comment by luc-github** [comment]
This commit (https://github.com/luc-github/Repetier-Firmware-0.92/commit/5fae938b383f36b13859680c85dcc49753b01b6e) allows to choose between Winstar LCD or HD44780 clone LCD in configuration.h even both seems working on my Duo 2SF, I have set the Winstar as Default Davinci one because we set 8bits communication and RW pin properly
I did not changed the logic for winstar so it is transparent change
As no more issues reported I close the issue

Thanks everyone for your great support
